Fix/#6821 import with transaction [EM - #6821]

spetros-do opened the pull request:
### Summary:
Completes the implementation of database transaction support for the emissions import process to prevent partial data imports when errors occur.

### Changes Made:
Ensure all database operations participate in single transaction
Implement rollback on any import error to prevent partial data
Update BulkLoadService and child service mocks with QueryRunner support
Add withTransaction utility function test coverage
Add comprehensive transaction mocking all test files

### Testing:
All test files now have proper transaction mocking
Child service mocks support transaction context propagation

yonatan-dp committed:
Complete Changes to Prevent Partial Imports on QA & EM

spetros-do committed:
AS part of the fixing of the partial imports of EM -- updated .spec file for "emissions-workspace" to ensure testing was not failing. The rest of the unit test passed with no issues.

spetros-do committed:
updated version of "@us-epa-camd/easey-common" from 21.1.10 to 21.1.16 to meet resolve conflict of ver. and additional test files that were in conflict during unit testing.

spetros-do committed:
updated version of "@us-epa-camd/easey-common" from 21.1.10 to 21.1.17 to meet resolve conflict of ver.

spetros-do committed:
Merge branch 'develop' into fix/#6821-import-with-transaction

mark-hayward-erg committed:
Merge branch 'develop' into fix/#6821-import-with-transaction

maxdiebold-erg commented on src/emissions-workspace/emissions.service.ts:
```diff
@@ -164,146 +165,154 @@ export class EmissionsWorkspaceService {
 
-    const importPromises = [
-      this.importDailyEmissions(
+      for (const monitorPlan of monitorPlans) {
```
Since all previous queries target unmodified data, it may be more clear to start the transaction here:
```suggestion
    const queryRunner = this.entityManager.connection.createQueryRunner();
    await queryRunner.startTransaction();
    
    try {
      const trx = queryRunner.manager;
      
      for (const monitorPlan of monitorPlans) {
```

spetros-do replied:
Thanks for the suggestion. I definitely see the performance benefits of moving the transaction boundary to only wrap data modifications.
However, my main concern is the small window where reference data (monitor plans, reporting periods, identifiers) could potentially change between validation and transaction start, potentially causing subtle consistency issues.
What do you think?
Based on your suggestion, here is how I would implement the change: 
```
// Pre-transaction validation phase - all read-only operations
    const stackPipeIds = objectValuesByKey<string>('stackPipeId', params, true);
    const unitIds = objectValuesByKey<string>('unitId', params, true);

    const plant = await this.plantRepository.getImportPlant({
      orisCode: params.orisCode,
      stackIds: stackPipeIds,
      unitIds: unitIds,
    });

    if (isUndefinedOrNull(plant)) {
      throw new NotFoundException('Plant not found.');
    }

    const monitorPlans = plant.monitorPlans;

    if (monitorPlans.length === 0) {
      throw new NotFoundException('Monitor plan not found.');
    }

    if (monitorPlans.length > 1) {
      throw new NotFoundException('Multiple active monitor plans found.');
    }

    const reportingPeriod = await this.entityManager.findOne(ReportingPeriod, {
      where: {
        year: params.year,
        quarter: params.quarter,
      },
    });

    if (!reportingPeriod) {
      throw new NotFoundException('Reporting period not found.');
    }

    const monitorPlanId = monitorPlans[0].id;
    const monitoringLocations = monitorPlans[0].locations;

    const reportingPeriodId = reportingPeriod.id;

    const identifiers = await this.getUnifiedIdentifiers(
      params,
      monitoringLocations,
      userId,
    );

    // Import-28 Valid formulaIdentifiers for location
    await this.checksService.invalidFormulasCheck(params, monitoringLocations);

    // Transaction phase - only data modification operations
    // Since all previous queries target unmodified data, it may be more clear to start the transaction here
    const queryRunner = this.entityManager.connection.createQueryRunner();
    await queryRunner.startTransaction();
```

maxdiebold-erg replied:
I think that's a valid concern, but we account for that by only letting one person check out and modify a configuration at a single time, so it should not be a problem.

I was thinking less about the performance and more about clarity and maintenance: currently, the transaction is started, then several queries are performed using non-transaction connections. The specific queries don't pose issues in this case, but moving the transaction boundary down would serve to say "hey, modifications are happening  and it's really only safe to use the transactional connection (`trx`) from here on out".

spetros-do replied:
Sounds good. Will adjust accordingly. Thanks.

maxdiebold-erg reviewed (dismissed):
Looks good! Left one nitpick suggestion but works well as is.

mark-hayward-erg commented on package.json:
```diff
@@ -38,7 +38,7 @@
     "@nestjs/typeorm": "^11.0.0",
     "@us-epa-camd/easey-common": "21.1.17",
-    "axios": "^1.11.0",
+    "axios": "^1.9.0",
```
This dependency downgrade shouldn't be included in the PR.

spetros-do replied:
Noted. Will make the change. Thx

spetros-do committed:
Merge remote-tracking branch 'origin/develop' into fix/#6821-em-import-with-transaction

spetros-do committed:
- Updated trx to start after validation phase for better performance.
- Updated test file to reflect new trx structure.

spetros-do committed:
Update package.json

spetros-do committed:
- Updated trx to start after validation phase for better performance.
- Updated test file to reflect new trx structure.

spetros-do committed:
Merge remote-tracking branch 'origin/fix/#6821-import-with-transaction' into fix/#6821-em-import-with-transaction

# Conflicts:
#	yarn.lock

maxdiebold-erg reviewed (dismissed):
Looks good, thanks for making that change!

spetros-do committed:
Resolved merge conflict in emissions.service.ts
- Updated trx to start after validation phase for better performance.
- Updated test file to reflect new trx structure.

mark-hayward-erg reviewed (changes requested):
This PR should not contain the manifest*.yml files, the package.json file, or the yarn.lock file.

By including them in this PR you will be undoing changes made on the develop branch that should not be affected by the functional changes for this ticket.

spetros-do commented:
I know...it was a merge conflict issue. Correcting. I am planning to remove it. 
Thx you though.

maxdiebold-erg commented:
All of the changes related to the `EarliestPartitionQuarter` should be added back, too: those changes were made a month ago as part of [#6770](https://github.com/US-EPA-CAMD/easey-ui/issues/6770).

spetros-do commented:
> All of the changes related to the `EarliestPartitionQuarter` should be added back, too: those changes were made a month ago as part of [#6770](https://github.com/US-EPA-CAMD/easey-ui/issues/6770).

Thx MadD. It was merge error as mentioned above. Will sort it out - but appreciate the feedback.

spetros-do commented:
Replaced by PR #613
